### PR TITLE
test interfaceObject

### PR DESCRIPTION
**Pull request by Geal**
*Description here*

Fixes #**issue_number**

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/tests/samples/basic/interface-object/plan.json [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            },
+            "expected_response": {
+                "data": {
+                    "req": {
+                        "id": "1",
+                        "c": "a"
+                    }
+                }
+            }
+        },
```
Comment (Geal) [inline apollo-router/tests/samples/basic/interface-object/plan.json:214]: this is an interesting failure. This is what it logs:

```
query: {"query":"query TestItf2 { req { id c } }"}

subgraphs received requests:
        method: POST
        path: http://localhost/
        {"query":"query TestItf2__accounts__0{req{__typename id i{__typename id x}}}","operationName":"TestItf2__accounts__0"}

        method: POST
        path: http://localhost/
        {"query":"query TestItf2__products__1($representations:[_Any!]!){_entities(representations:$representations){...on I{y}}}","operationName":"TestItf2__products__1","variables":{"representations":[{"__typename":"I","id":"1"}]}}

assertion `left == right` failed
expected: {"data":{"req":{"id":"1","c":"a"}}}
received: {"data":{"req":{"id":"1","c":null}},"errors":[{"message":"HTTP fetch failed from 'reviews': 400: Bad Request","path":["req"],"extensions":{"code":"SUBREQUEST_HTTP_ERROR","service":"reviews","reason":"400: Bad Request","http":{"status":400}}},{"message":"Unknown type \"C\".","path":["req"],"extensions":{"code":"GRAPHQL_VALIDATION_FAILED"}}]}
```

The `reviews` subgraph should have received the query `query TestItf2__reviews__2($representations:[_Any!]!){_entities(representations:$representations){...on C{c}}}` with variables `{"representations": [{"__typename": "C"), "i": {"x": 1, "y": null}, "id": "1"}]}`, so this means that something is validating graphql queries between the moment the subgraph request is created, and the moment it reaches the subgraph

Comment (Geal) [reply apollo-router/tests/samples/basic/interface-object/plan.json:214]: this is the subgraph schema that was extracted by the planner for `reviews`:

```graphql
schema
  @link(url: "https://specs.apollo.dev/link/v1.0")
  @link(url: "https://specs.apollo.dev/federation/v2.9", import: ["@key", "@requires", "@provides", "@external", "@tag", "@extends", "@shareable", "@inaccessible", "@override", "@composeDirective", "@interfaceObject"])
{
  query: Query
}

directive @link(url: String, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA

directive @key(fields: federation__FieldSet!, resolvable: Boolean = true) repeatable on OBJECT | INTERFACE

directive @requires(fields: federation__FieldSet!) on FIELD_DEFINITION

directive @provides(fields: federation__FieldSet!) on FIELD_DEFINITION

directive @external(reason: String) on OBJECT | FIELD_DEFINITION

directive @tag(name: String!) repeatable on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION | SCHEMA

directive @extends on OBJECT | INTERFACE

directive @shareable repeatable on OBJECT | FIELD_DEFINITION

directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION

directive @override(from: String!, label: String) on FIELD_DEFINITION

directive @composeDirective(name: String) repeatable on SCHEMA

directive @interfaceObject on OBJECT

directive @federation__authenticated on FIELD_DEFINITION | OBJECT | INTERFACE | SCALAR | ENUM

directive @federation__requiresScopes(scopes: [[federation__Scope!]!]!) on FIELD_DEFINITION | OBJECT | INTERFACE | SCALAR | ENUM

directive @federation__policy(policies: [[federation__Policy!]!]!) on FIELD_DEFINITION | OBJECT | INTERFACE | SCALAR | ENUM

directive @federation__sourceAPI repeatable on SCHEMA

directive @federation__sourceType repeatable on OBJECT | INTERFACE

directive @federation__sourceField repeatable on FIELD_DEFINITION

directive @federation__context(name: String!) repeatable on INTERFACE | OBJECT | UNION

directive @federation__fromContext(field: federation__ContextFieldValue) on ARGUMENT_DEFINITION

directive @federation__cost(weight: Int!) on ARGUMENT_DEFINITION | ENUM | FIELD_DEFINITION | INPUT_FIELD_DEFINITION | OBJECT | SCALAR

directive @federation__listSize(assumedSize: Int, slicingArguments: [String!], sizedFields: [String!], requireOneSlicingArgument: Boolean = true) on FIELD_DEFINITION

enum link__Purpose {
  """
  `SECURITY` features provide metadata necessary to securely resolve fields.
  """
  SECURITY

  """
  `EXECUTION` features provide metadata necessary for operation execution.
  """
  EXECUTION
}

scalar link__Import

scalar federation__FieldSet

scalar federation__Scope

scalar federation__Policy

scalar federation__ContextFieldValue

interface I {
  id: ID!
  x: Int
  y: Int
}

type C
  @key(fields: "id")
{
  id: ID! @shareable
  i: I @external
  c: String @requires(fields: "i { x y }")
}

scalar _Any

type _Service {
  sdl: String
}

union _Entity = C

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
}
```

Comment (Geal) [reply apollo-router/tests/samples/basic/interface-object/plan.json:214]: this ws an issue in the way I wrote mocks, the request was actually going to the starstuff subgraph

File: apollo-router/tests/samples/basic/interface-object/plan.json [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+            },
+            "expected_response": {
+                "data": {
+                    "req": {
+                        "id": "1",
+                        "c": "c"
+                    }
+                }
+            }
+        },
```
Comment (Geal) [inline apollo-router/tests/samples/basic/interface-object/plan.json:249]: this is now failing with the following log:
```
query: {"query":"query TestItf2 { req { id c } }"}

subgraphs received requests:
        method: POST
        path: http://localhost/
        {"query":"query TestItf2__accounts__0{req{__typename id i{__typename id x}}}","operationName":"TestItf2__accounts__0"}

        method: POST
        path: http://localhost/
        {"query":"query TestItf2__products__1($representations:[_Any!]!){_entities(representations:$representations){...on I{y}}}","operationName":"TestItf2__products__1","variables":{"representations":[{"__typename":"I","id":"1"}]}}

        method: POST
        path: http://localhost/
        {"query":"query TestItf2__reviews__2($representations:[_Any!]!){_entities(representations:$representations){...on C{c}}}","operationName":"TestItf2__reviews__2","variables":{"representations":[{"__typename":"C","i":{"x":1,"y":null},"id":"1"}]}}

assertion `left == right` failed
expected: {"data":{"req":{"id":"1","c":"c"}}}
received: {"data":{"req":{"id":"1","c":null}},"errors":[{"message":"HTTP fetch failed from 'reviews': 404: Not Found","path":["req"],"extensions":{"code":"SUBREQUEST_HTTP_ERROR","service":"reviews","reason":"404: Not Found","http":{"status":404}}},{"message":"HTTP fetch failed from 'reviews': subgraph response does not contain 'content-type' header; expected content-type: application/json or content-type: application/graphql-response+json","path":["req"],"extensions":{"code":"SUBREQUEST_HTTP_ERROR","service":"reviews","reason":"subgraph response does not contain 'content-type' header; expected content-type: application/json or content-type: application/graphql-response+json","http":{"status":404}}}]}
```

The request to the reviews subgraph should have the following variables: `"representations":[{"__typename":"C","i":{"x":1,"y":1},"id":"1"}]}` but `y` is null.

this is the content of the fetch node's `requires` field: `[InlineFragment(InlineFragment { type_condition: Some(\"C\"), selections: [Field(Field { alias: None, name: \"__typename\", selections: None }), Field(Field { alias: None, name: \"i\", selections: Some([Field(Field { alias: None, name: \"x\", selections: None }), Field(Field { alias: None, name: \"y\", selections: None })]) }), Field(Field { alias: None, name: \"id\", selections: None })] })]` 
and the available data when trying to extract variables:
`{\"req\":{\"__typename\":\"C\",\"id\":\"1\",\"i\":{\"__typename\":\"A\",\"id\":\"1\",\"x\":1}}}`

Comment (Geal) [reply apollo-router/tests/samples/basic/interface-object/plan.json:249]: this is the generated query plan:
```
QueryPlan {
  Sequence {
    Fetch(service: "accounts") {
      {
        req {
          __typename
          id
          i {
            __typename
            id
            x
          }
        }
      }
    },
    Flatten(path: "req.i") {
      Fetch(service: "products") {
        {
          ... on I {
            __typename
            id
          }
        } =>
        {
          ... on I {
            y
          }
        }
      },
    },
    Flatten(path: "req") {
      Fetch(service: "reviews") {
        {
          ... on C {
            __typename
            i {
              x
              y
            }
            id
          }
        } =>
        {
          ... on C {
            c
          }
        }
      },
    },
  },
}
```

Comment (Geal) [reply apollo-router/tests/samples/basic/interface-object/plan.json:249]: this was an issue in mocking too

**Review by bnjjj — APPROVED** [review]

**Pull request merged**